design/design-doc-template-stacks-experience.md

```diff
@@ -0,0 +1,473 @@
+for the template rendered by the controller. For more details, see the
+[detailed examples](https://github.com/suskin/template-stack-experience)
+or the [design document about the
+internals](https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-template-stacks.md).
+
+### Templating/configuration engine
+
+Template stacks will not be opinionated about which templating engine is
+used. We plan to support multiple configuration engines. The engine will
+be configurable by setting values in the `stack.yaml`.
+
+### Lifecycle hooks
+
+We expect to eventually support lifecycle hooks. See the [speculative
+design in the template stacks experience repo](https://github.com/suskin/template-stack-experience/tree/master/wordpress-workload/go-templates/lifecycle-hook)
+for more details about what that could look like.
+
+### Internal representation of templates
+
+See the [design doc on the internals of the Template Stack
+implementation](https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-template-stacks.md).
+
+### The template stack controller
+
+See the [design doc on the internals of the Template Stack
+implementation](https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-template-stacks.md).
+
+## Example use-cases
+
+See the [Template Stack
+Experience](https://github.com/suskin/template-stack-experience) code
+examples for some realistic examples of how different use-cases work
+with a realistic workload.
+
+Examples covered by the repository all currently use go templates as the
+templating engine. The following cases are covered by the examples:
+
+* No variables at all in a set of configurations.
+* Variables in configurations with default variable values.
+* Variables in configurations, with values specified by users when a
+  render is requested.
+* Resource packs.
+* Multiple CRDs defined in a single Stack.
+* Multiple rendered variants of a given configuration.
+* Composing configuration using multiple Stacks.
+
+### Helm charts
+
+To author a template stack from a simple helm chart, these steps could
+be followed:
+
+1. Create and initialize a template stack project, with a helm flag:
+   `kubectl crossplane stack init --template`.
+2. Create a CRD: `kubectl crossplane stack crd init WordpressInstance
+   wordpressinstances wordpress.samples.stacks.crossplane.io`
+3. Put the chart's contents into the templates folder.
+4. Create a `stack.yaml` in the root of the repostory, and configure it
+   to use the templates for the created CRD.
+5. To set up default values:
+    1. Put the contents of the chart's `values.yaml` into the CRD's
+       [default field value definitions][kubernetes-default-values].
+6. Bundle and publish the stack.
+
+When consuming the stack, the default template values can be overridden
+by specifying fields on the render request object.
+
+This use-case will probably need additional thought if we want to
```

Comment:
It might be worth mentioning some helm chart differences one may encounter when converting a helm  chart to fit this initial template stack engine (or we can add this after the implementation).

* includes may not initially be supported but are planned, file paths may need to change.
* template variables may need to be altered
* template functions may not align with helm

It may also be worth pointing out that a future template stacks engine could potentially consume helm charts with less manipulation.

Reply:
why would there be a conversion necessary of a helm chart to fit into a stack? if we are reusing the helm rendering engine under the covers, similar to how we'd reuse any existing engine for existing formats we want to support, why would there be any differences that a stack author would have to learn and deal with?

Reply:
> converting a helm chart to fit this initial template stack engine 

I have been expecting that the initial engine would not use helm directly under the covers, but would provide helm-like functionality through the use of Go templates and functions similar when not identical to those offered by helm.

If the first engine or any subsequent engine does employ Helm directly, I still expect there to be differences as helm runs locally against a local environment and filesystem and maintains some local state, while the Template Stacks experience of this would be fundamentally different.

Whether there are limitations (on customizations or functions available, versioning, channel upgrades and rollbacks), new features, caveats, conversion steps, or work-arounds for Helm users coming to Template Stacks, I would expect a brief explanation of the expected differences.  For example, fields outside of those defined in Helm's values.yaml can be updated, similar to how Kustomize can overlay values.  Will the Helm Template Stack Engine offer that capability?  Would an arbitrary yaml field be the interface? Is Tiller used and/or required? If not how does this change the Helm user experience?

Reply:
I'm definitely thinking that reusing existing engines (such as helm but there are others too) instead of writing our own implementation would be the fastest way to bootstrap this effort and provide the most functionality at the same time. Agreed that there could be differences that come up in reality and it would be good to understand them.

Reply:
> why would there be a conversion necessary

It is very difficult to be confident _a priori_ that something that works in one situation will work in a slightly different situation where a different code path is being executed. In this case, we would be going from using `helm install` to using `helm template`, which is not the same code path and already has some known other differences:

* `helm template` does not use tiller
* From the helm template documentation, `any values that would normally be
looked up or retrieved in-cluster will be faked locally.`

So, while we may not know of any obviously necessary conversion steps yet, that doesn't mean there won't be any.

> Kustomize. Will the Helm Template Stack Engine offer that capability?

Based on my understanding, I believe using kustomize after helm still would require a kustomization file, but the short answer is that this should be possible. Since we're planning on mostly decoupling the processing engine from the stack itself, "helm to kustomize" could just be another engine that is supported.